### DISCUSSION TITLE
SVG Rendering

### DISCUSSION THREAD

**Discussion by marchbold** [airsdk/Adobe-Runtime-Support#2779] [Q&A]
Anyone know of a good SVG Rendering library? I need to add rendering of svgs to an desktop flex / air app 

I've tried an old version of the svgweb toolkit that I had which works for simple svgs but unfortunately it doesn't seem to support some of the more modern features that I need in this case.

**Comment by marchbold** [comment]
(seems I've already asked this same question https://github.com/airsdk/Adobe-Runtime-Support/discussions/697 but if anyone has done anything new would be very interested)

**Comment by RayGyoe** [comment]
I used this library to encapsulate ANE when loading SVG on the desktop（ https://github.com/sammycage/lunasvg ）

**Reply by marchbold** [reply]
Ah cool, unfortunately I was hoping to stay away from ANEs with this one. They distribute an `air` package currently and was hoping not to have to change their delivery.

**Reply by ajwfrost** [reply]
If it's pure C++, could you use Crossbridge and turn it into a SWC..?

**Reply by marchbold** [reply]
Ooo totally forgot about that, thanks for the reminder! Will investigate!

**Reply by raresn** [reply]
This would be great! We now have assets in 3x sizes and package them up and choose what to use when the app starts. This will be super useful to just put everything in 1 svg and scale at the beginning :)

**Reply by raresn** [reply]
@marchbold is this a solution you plan to release to the public as well? :) ( just thought about the fact that it might be private )

**Reply by marchbold** [reply]
I'll discuss with my client but most likely anything we do here I'll be able to opensource so should be a public solution.

**Reply by RayGyoe** [reply]
https://github.com/airsdk/Adobe-Runtime-Support/assets/336440/b436b57f-d2b2-41b5-b055-b71ac78e271a

Took some time to integrate into ANEWinCore

**Reply by marchbold** [reply]
@RayGyoe Unfortunately an ANE is not an option for me here.

**Reply by RayGyoe** [reply]
Keep a file for others to search for ~ 😁

**Comment by raresn** [comment]
Would this work on ios/android as well? Would be supercool to have a svg with all the assets in 1 size and scale it on startup based on the screen size then use it like that...

**Comment by bwhiting** [comment]
What about a pure as3 version?
https://github.com/lucaslorentz/AS3SVGRenderer/tree/master/SVGRenderer
Saw that as a first result but no idea how good it is or how complete you need it to be. Am sure there are other libs out there too.

**Reply by marchbold** [reply]
Yes I have tried that one, but unfortunately it seems to have some issues with detecting the size of the svg, I had better luck     https://code.google.com/archive/p/svgweb/ and have used that in the past but it doesn't support more modern features which I need currently.

**Comment by 2jfw** [comment]
Would be nice to have SVG support natively in AIR - any chance for adding this in a future version maybe @ajwfrost ?
This seems quite similar to the support of more "contemporary" video codecs (H265 or VP9 etc..) - positive side effect would be that this could boost the acceptance of AIR and make it be considered more often for creating apps etc..

**Comment by xabip** [comment]
We use a webview on top of the display list to show the SVGs

**Reply by marchbold** [reply]
Yeah that is an option I am exploring as well, just this project has quite a complex mxml component layout and positioning the webviews has proved problematic in the past.

**Comment by civet** [comment]
Recently I tried to fix an old [AS3 SVG library](https://github.com/civet/svg-parser).  It can be used to display some simple SVGs, although it still has many issues. In fact, its code is not complicated, and it is easy to extend it and implement more features. PRs are welcome.

Flash/AIR natively supports vector drawing and E4X, which is very suitable for parsing and rendering SVG.

**Reply by marchbold** [reply]
Thanks, unfortunately that one doesn't render the SVG I need currently either. I've ended up using a webview and drawing that to a bitmap. Working well enough until I get more time to look into crossbridge again.

**Reply by pol2095** [reply]
I tried to implement [lunasvg](https://github.com/sammycage/lunasvg ) using crossbridge [here](http://pol2095.free.fr/SVG2Bitmap/)

**Reply by marchbold** [reply]
@pol2095 Did you get it to work? I haven't had much time to look at it yet, having issues getting crossbridge to work on Apple silicon. Seems it doesn't support the new Apple c compiler with an architecture flag.

**Reply by pol2095** [reply]
I tried a few SVG, it work (I used a Mac Intel to create the swc).

**Reply by pol2095** [reply]
I creatad a worker version [here](http://pol2095.free.fr/SVG2Bitmap/SVG2BitmapWorker/)

**Reply by Devin-Licastro** [reply]
Hey @marchbold I might need to implement some SVG rendering soon. For the solution you found, can I test my understanding, is it like this?

1. On the screen where you need to display the SVG, momentarily make everything invisible
2. Open a webview and load the svg... actually I don't quite understand how you get the svg in the webview as I've only opened webpages with them before. Also are you using the distriqt webview ANE for this?
3. Render the contents of the webview to a bitmap 
4. Hide the webview, make everything else visible again, load and display the bitmap 

Is that the general idea?

**Reply by marchbold** [reply]
Hi, I had the requirement of not using any ANEs (due to being distributed as an air file) so I just used StageWebView. 

1. I created a `StageWebView` instance and placed it "off screen"

```as3
var _webView:StageWebView;

_webView = new StageWebView();
_webView.viewPort = new Rectangle( -200, -200, 200, 200 );
_webView.addEventListener( Event.COMPLETE, webview_handleLoadComplete );
_webView.stage = stage;
_webView.loadString( getHtmlForUrl( url, 200 ) );
```

2. I was loading the svg from a url so I created some html to display the svg at a set width:

```as3
private static const HTML_TEMPLATE:String = <html>
			<object data="%SVGURL" type="image/svg+xml" width="%WIDTH"></object>
		</html>.toString();


private function getHtmlForUrl( svgUrl:String, width:int = 200 ):String
{
	return HTML_TEMPLATE.replace( "%SVGURL", svgUrl ).replace( "%WIDTH", width );
}
```

3. On complete I rendered the webview to a bitmap and then used that in the app:

```as3
var _svgBitmap:Bitmap;

function webview_handleLoadComplete( event:Event ):void
{
	var bitmapData:BitmapData = new BitmapData( _webView.viewPort.width, _webView.viewPort.height );
	_webView.drawViewPortToBitmapData( bitmapData );
	_webView.stage = null;

	_svgBitmap = new Bitmap( bitmapData );
	addChild( _svgBitmap );
}
```

**Reply by Devin-Licastro** [reply]
Thanks a bunch for the detailed reply!

**Reply by Ender22** [reply]
Hey @marchbold it's been a while, I'm wondering if this is still the best way to render svg's?
I actually need to render them in Starling for iOS, Android, Windows and Mac, from what I've seen the above method might only work on mobile. 
I was thinking maybe I coud use the NativeWebView ANE, but then I believe I wouldnt be able to draw its viewport to bitmap because it's on a higher layer? 
Any tips here?

**Reply by Ender22** [reply]
ooh, I just found the drawToBitmap support on the webview, I'll give that a try! 
https://docs.airnativeextensions.com/docs/nativewebview/screenshot

**Reply by pol2095** [reply]
@Ender22 are-you tried crossbridge implementation of LunaSVG ?

**Reply by Ender22** [reply]
I have not, would that be this one here? 
https://github.com/sammycage/lunasvg
I can give it a shot

**Reply by raresn** [reply]
@Ender22 we are also trying to render svgs at different resolutions. Please let me know if you find a solution :)

**Reply by pol2095** [reply]
@Ender22 yes, crossbridge swc use [https://github.com/sammycage/lunasvg](https://github.com/sammycage/lunasvg).

**Reply by marchbold** [reply]
@Ender22 Yeah the webview rendering to bitmap data was what I ended up with but I haven't investigated it again since this original post. 

A crossbridge implementation would have been perfect but my client didn't have the time / resources available at the time to look into that approach. If I get some time after christmas I'll have a look into it. It's still something that comes up often enough for me.

**Reply by Ender22** [reply]
@pol2095 thanks. Umm would you mind spelling it out for me like I'm 5 what i would do to try this? I have not heard of crossbridge before this.

Is the idea to download this library https://github.com/sammycage/lunasvg then use crossbridge to convert it to as3 classes or swc or something and then I'd be able to use it? Or would crossbridge like compile it in some way like how ANEs work? 

Thanks for any info

**Reply by pol2095** [reply]
crossbridge convert C/C++11 code to as3 bytecode, use just the swc.

**Comment by Fancy2209** [comment]
I compiled the [OpenFL SVG](https://github.com/openfl/svg) Library into a SWC for AS3: https://github.com/Fancy2209/OpenFL-SVG-on-AS3-Test/raw/refs/heads/main/libs/svg.swc

It can be used like this, where svgData is a String containing the contents of an SVG file.
```actionscript
import format.SVG;
var svgDO:Shape = new Shape();
var svgRenderer:SVG = new SVG(svgData);
svgRenderer.render(svgDO.graphics);
```

Here's an example project using it and FeathersUI OpenFL on AS3: https://github.com/Fancy2209/OpenFL-SVG-on-AS3-Test/

**Reply by marchbold** [reply]
Love it!!
